### PR TITLE
Custom ciph constants

### PR DESCRIPTION
**Pull request by beldmit**
Here is a set of custom values to be used in the ctrl functions of non-core implementations of the cipher algorithms.

### CODE REVIEW

**Comment by levitte** [comment]
I think such custom names are quite useless, unless the caller knows for sure what implementation of a certain cipher it's dealing with.  Have descriptive names for their intended use instead, preferably named in such a way that other (possibly future) ciphers can reuse them for the same purpose.

The next trick is to figure out a general purpose enough structure for the parameters that should be sent through the control calls.

**Comment by beldmit** [comment]
The previous version of the patch was:

+/*
+ * Russian GOST has some parameters defining its usage:
+ * S-blocks, key meshing, padding modes
+ */
+#define          EVP_CTRL_GOST_PARAMS       0x1d
+#define          EVP_CTRL_GOST_KEY_MESHING  0x1e
+#define          EVP_CTRL_GOST_PADDING      0x1f
+

I am not sure there is any universal purpose (except padding).

**Comment by levitte** [comment]
The previous version of the patch was:

> +/*
> * Russian GOST has some parameters defining its usage:
> * S-blocks, key meshing, padding modes
> */
> +#define EVP_CTRL_GOST_PARAMS 0x1d
> +#define EVP_CTRL_GOST_KEY_MESHING 0x1e 
> +#define EVP_CTRL_GOST_PADDING 0x1f +
> I am not sure there is any universal purpose (except padding).

Hmmm, although I don't really like having implementation specific names like that, it was still better, in my opinion.

File: include/openssl/evp.h [hunk 1]
```diff
@@ -427,6 +427,19 @@ struct evp_cipher_st {
...
+/* EVP_CTRL_SET_SBOX takes the char* specifying S-boxes */
+# define         EVP_CTRL_SET_SBOX                       0x1d
+/* 
+ * EVP_CTRL_SBOX_USED takes a 'size_t' and 'char *', pointing at a
+ * pre-allocated buffer with specified size 
+ */
+# define         EVP_CTRL_SBOX_USED                      0x1e
+/* EVP_CTRL_KEY_MESH takes 'size_t' number of bytes to mesh the key after, 0 switches meshing off */
+# define         EVP_CTRL_KEY_MESH                       0x1f
+/* EVP_CTRL_BLOCK_PADDING_MODE takes the char* specifying padding mode */
```
Comment (mattcaswell) [inline include/openssl/evp.h]: Question from Steve "Why does this take char * for the mode?"

Comment (beldmit) [inline include/openssl/evp.h]: It seems to be a simple solution. Otherwise we can specify a set of constants and pass them via the int parameter.

Comment (mattcaswell) [inline include/openssl/evp.h]: What values do you currently accept?

Comment (beldmit) [inline include/openssl/evp.h]: ISO/IEC 7816-4
Zero padding
PKCS#7 padding

Comment (levitte) [inline include/openssl/evp.h]: Speaking from a general point of view, I think specifying them as strings is a good thing. It avoids the hassle of having to define symbols or the use of anonymous numbers, and truly becomes a communication between the app and the underlying algo with none else (OpenSSL itself included)  being the wiser.

**Comment by mattcaswell** [comment]
This has now been merged (finally!). Thank you for your patience. :-)

**Pull request closed without merging**